### PR TITLE
PLANNER-2416: Assign a time budget for the LS to execute

### PR DESCRIPTION
**Pull request by wmedvede**
Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [X] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [X] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [X] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [X] Pull Request contains link to the JIRA issue
- [X] Pull Request contains link to any dependent or related Pull Request
- [X] Pull Request contains description of the issue
- [X ] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Pull Request</b>  
  Please add comment: <b>Jenkins retest this</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

</details>

### CODE REVIEW

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
Hi @ge0ffrey @rsynek would you mind review this PR?
Thanks

File: task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/config/TaskAssigningConfigValidator.java [hunk 1]
```diff
@@ -77,4 +83,20 @@ private static void validateUserSyncConfig(TaskAssigningConfig config) {
...
+                config.getWaitForImprovedSolutionDuration());
+    }
+
+    private static void validateImproveSolutionOnBackgroundDuration(TaskAssigningConfig config) {
+        validatePropertyIsZeroOrGreaterThanZeroMillisecondsDuration(TaskAssigningConfigProperties.IMPROVE_SOLUTION_ON_BACKGROUND_DURATION,
+                config.getImproveSolutionOnBackgroundDuration());
+    }
+
+    private static void validatePropertyIsZeroOrGreaterThanZeroMillisecondsDuration(String propertyName, Duration duration) {
+        if (duration.isNegative() || (!duration.isZero() && duration.toMillis() < 1)) {
```
Comment (rsynek) [inline task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/config/TaskAssigningConfigValidator.java:98]: The `duration.isZero()` check seems to be redundant. Also, are two millis good enough? Are we able to react so quickly?

File: task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java [hunk 2]
```diff
@@ -308,29 +342,63 @@ void onBestSolutionChange(BestSolutionChangedEvent<TaskAssigningSolution> event)
...
+    private void scheduleOnBestSolutionChange(TaskAssigningSolution chBestSolution, Duration duration) {
+        if (!isWaitForImprovedSolutionTimerScheduled() && !context.isCurrentChangeSetProcessed()) {
+            LOGGER.debug("Schedule execute solution change with waiting duration: {}", duration);
+            scheduleWaitForImprovedSolutionTimer(duration, chBestSolution);
+        }
+    }
+
+    private void executeSolutionChange(TaskAssigningSolution chBestSolution, Supplier<TaskAssigningSolution> solutionSupplier) {
+        TaskAssigningSolution currentLastBestSolution = solutionSupplier.get();
+        LOGGER.debug("Executing delayed solution change for currentChangeSetId: {}, chBestSolution: {}, lastBestSolution: {}",
```
Comment (rsynek) [inline task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java]: I doubt anybody can recognize what `chBestSolution` is; if the debugging line stays, it should be more descriptive (repeats in other debug lines, too).

File: task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java [hunk 3]
```diff
@@ -274,18 +281,45 @@ void processDataEvents(List<DataEvent<?>> events) {
...
                 }
             }
         } catch (Exception e) {
             failFast(e);
-        } finally {
-            lock.unlock();
         }
     }
 
+    private boolean hasToApplyImprovedOnBackgroundSolution(SolutionUpdatedOnBackgroundDataEvent dataEvent,
```
Comment (rsynek) [inline task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java:306]: `dataEvent` parameter name has been probably derived from its supertype; could we use a more specific name?

Comment (rsynek) [inline task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java:306]: Maybe `hasSolutionImprovedOnBackground` name would fit the method better? It informs whether the solution has improved or not; decision whether it should be applied is out of its scope.

File: task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java [hunk 4]
```diff
@@ -351,17 +419,27 @@ private void executeSolutionChange(TaskAssigningSolution solution) {
...
 
+    void onSolutionImprovedEvent(@Observes SolutionImprovedEvent event) {
+        LOGGER.debug("onSolutionImprovedEvent: timerId: {}", event.getTimerId());
+        executeSolutionChange(event.getChBestSolution(), lastBestSolution::get);
+    }
+
+    void onSolutionImprovedOnBackgroundEvent(@Observes SolutionImprovedOnBackgroundEvent event) {
+        LOGGER.debug("onSolutionImprovedOnBackgroundEvent: timerId: {}", event.getTimerId());
+        serviceEventConsumer.accept(new SolutionUpdatedOnBackgroundDataEvent(event.getTimerId(), ZonedDateTime.now()));
+    }
```
Comment (rsynek) [inline task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java:441]: @wmedvede Do I understand these two events correctly?
`SolutionImprovedEvent` : fired by a timer; effectively is a delay between the last `BestSolutionChangedEvent` from OptaPlanner and its propagation outside the task assigning.
`SolutionImprovedOnBackgroundEvent` : fired by a different timer; it's how long are we willing to wait for OptaPlanner to produce some better solution?
It's hard for me to see a difference between these two events.

File: task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/config/TaskAssigningConfigValidator.java (hunk 1 above)
Comment (wmedvede) [reply task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/config/TaskAssigningConfigValidator.java:98]: This check is really to be sure that only values supported by the underlying timers mechanism are introduced.
A lower boundary of 5 mills, 10 millis, 100 millis is something completely empirically that I won't introduce. So yeah, it'll be up to the user to configure a value that make sense. If a value o two milliseconds is introduced, the effect is that the probability of getting an improved solution in this period will be very low.

File: task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java (hunk 4 above)
Comment (wmedvede) [reply task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java:441]: SolutionImprovedEvent: I'ts the time we let the solver improve a solution before we use it.  Changes -> initial CH Solution -> (wait some time and let the LS execute) -> SolutionImprovedEvent -> we can use the solution.

SolutionImprovedOnBackgroundEvent: It's a configurable amount of time that we continue listening for solutions (after executing the last one) to see if this solution is even better. In that case, if potential planning items are derived from it, we apply them.

Changes -> initial CH Solution -> (wait optionally for the LS to execute) -> we apply the corresponding planning items -> (wait some time while listening for LS produced solutions) -> SolutionImprovedOnBackgroundEvent -> Do we have a better solution? Do we have any planning items that could make sense to execute?  If yes, apply them.

File: task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/config/TaskAssigningConfigValidator.java (hunk 1 above)
Comment (wmedvede) [reply task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/config/TaskAssigningConfigValidator.java:98]: btw, the duration.isZero() is not redundant.

File: task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java (hunk 3 above)
Comment (wmedvede) [reply task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java:306]: not apply.

Comment (wmedvede) [reply task-assigning/task-assigning-service/src/main/java/org/kie/kogito/taskassigning/service/TaskAssigningService.java:306]: I think it's ok.

**Comment by wmedvede** [comment]
Jenkins please retest this

**Comment by wmedvede** [comment]
@rsynek suggestions has been applied

**Review by rsynek — APPROVED** [review]

**Pull request merged**